### ISSUE TITLE
How to split my dataset into a train and test data but now, using cross_validate()

### ISSUE THREAD

**Issue by NaveganteX** [NicolasHug/Surprise#277]
#### Description

Hello, Nicolas!

Hope all is good! I am doing my master thesis and I am using your amazing Surprise package, which you made for the occasion (PhD thesis).

I wanted to ask you how can I use your cross_validate() class instead of GridSearch(SVD, param_grid, measures=['RMSE'], verbose=0).

I want to set cross_validate(whatever_algo,cv=10,n_jobs=-1). I have attached your original solution below. I added what I wanted (commented out, obviously)

Thank you very much in advance!

#### Steps/Code to Reproduce
                            
                                                                           
```
from __future__ import (absolute_import, division, print_function,         
                        unicode_literals)      
import random                                                              
                                                                           
from surprise import SVD                                                   
from surprise import Dataset                                               
from surprise import accuracy                                              
from surprise import GridSearch                                            
                                                                           
                                                                           
# Load your full dataset.                                                  
data = Dataset.load_builtin('ml-100k')                                     
raw_ratings = data.raw_ratings                                             
                                                                           
# shuffle ratings if you want                                              
random.shuffle(raw_ratings)                                                
                                                                           
# 90% trainset, 10% testset                                                
threshold = int(.9 * len(raw_ratings))                                     
trainset_raw_ratings = raw_ratings[:threshold]                             
test_raw_ratings = raw_ratings[threshold:]                                 
                                                                           
data.raw_ratings = trainset_raw_ratings  # data is now your trainset       
data.split(n_folds=3)                                                      
                                                                           
 
print('GRID SEARCH...')                                                    
param_grid = {'n_epochs': [5, 10], 'lr_all': [0.002, 0.005]}   

# Instead of GridSearch() use cross_validate()
# cross_validate =cross_validate(SVD(n_epochs = i,n_factors=10, biased=True), data_svd, measures=['rmse'], cv=10, verbose=True,n_jobs=-1)
            
grid_search = GridSearch(SVD, param_grid, measures=['RMSE'], verbose=0)    
grid_search.evaluate(data)                                                 
                                                                           
algo = grid_search.best_estimator['RMSE']                                  
                                                                           
# retrain on the whole train set                                           
trainset = data.build_full_trainset()                                      
algo.train(trainset)                                                       
                                                                           
# now test on the trainset                                                 
testset = data.construct_testset(trainset_raw_ratings)                     
predictions = algo.test(testset)                                           
print('Accuracy on the trainset:')                                         
accuracy.rmse(predictions)                 

```

**Comment by NicolasHug** [comment]
The doc is here https://surprise.readthedocs.io/en/stable/model_selection.html#cross-validation

**Comment by NaveganteX** [comment]
`cross_validate()` returns a dict. I will try to do my best and I will be back if I cannot make it work!

**Comment by NaveganteX** [comment]
This is what I was able to do. I couldn't use cross_validate(), instead I used GridSearchCV() with cv = 10.

Do you think it is correct?

```
from __future__ import (absolute_import, division, print_function,         
                        unicode_literals)                                  
                                                                           
import random                                                              
                                                                           
from surprise import SVD                                                   
from surprise import Dataset                                               
from surprise import accuracy                                              
from surprise import GridSearch
from surprise.model_selection import GridSearchCV

                                                                           
                                                                           
# Load your full dataset.                                                  
data = Dataset.load_builtin('ml-100k')                                     
raw_ratings = data.raw_ratings                                             
                                                                           
# shuffle ratings if you want                                              
random.shuffle(raw_ratings)                                                
                                                                           
# 90% trainset, 10% testset                                                
threshold = int(.9 * len(raw_ratings))                                     
trainset_raw_ratings = raw_ratings[:threshold]                             
test_raw_ratings = raw_ratings[threshold:]                                 
                                                                           
data.raw_ratings = trainset_raw_ratings  # data is now your trainset       
data.split(n_folds=3)                                                      
                                                                           
# Select your best algo with grid search. Verbosity is buggy, I'll fix it. 
print('GRID SEARCH...')                                                    
param_grid = {'n_epochs': [5, 10], 'lr_all': [0.002, 0.005]}
grid_search = GridSearchCV(SVD, param_grid,cv=10, measures=['RMSE'],n_jobs=-1)    
#algo =cross_validate(SVD(n_epochs = i,n_factors=10, biased=True), data_svd, measures=['rmse'], cv=10,n_jobs=-1)
#grid_search.evaluate(data)                                                 

grid_search.fit(data)                                                                           
algo=grid_search.best_estimator['rmse']    


                                                                           
# retrain on the whole train set                                           
trainset = data.build_full_trainset()                                      
algo.train(trainset)                                                       
                                                                           
# now test on the trainset                                                 
testset = data.construct_testset(trainset_raw_ratings)                     
predictions = algo.test(testset)                                           
print('Accuracy on the trainset:')                                         
accuracy.rmse(predictions)                                                 
                                                                           
# now test on the testset                                                  
testset = data.construct_testset(test_raw_ratings)                         
predictions = algo.test(testset)                                           
print('Accuracy on the testset:')                                          
accuracy.rmse(predictions) 

print(grid_search.best_params['rmse'])

```

**Comment by NicolasHug** [comment]
Sorry I don't understand what you're trying to do.

There isn't much more to `cross_validate` than what's already in the doc (and the user guid which is linked to the doc)

**Comment by NaveganteX** [comment]
I have the training and test sets on separate files.

Then, I would like to use cross-validation (`cross_validate()`) using only the training set, so I can tune the parameters of the algorithm.

Finally, I would like to test the performance of the best algorithm using the test set which is a different one but of course, it has the same structure i.e userid, movieid, rating.

**Comment by NicolasHug** [comment]
https://surprise.readthedocs.io/en/stable/FAQ.html#how-to-save-some-data-for-unbiased-accuracy-estimation

**Comment by NaveganteX** [comment]
Thank you very much, Nicolas!  I really appreciate it!
